File: 2015/Day04_2015.py
import hashlib

secretKey = "ckczppom"
keyFollower = 1

while True:
    inputKey = secretKey + str(keyFollower)
    hashedKey = hashlib.md5(inputKey.encode())

    if hashedKey.hexdigest()[0:5] == "00000":
        print("The lowest positive number leading to the MD5 hash starting with at least 5 zeroes is: ", end = "")
        print(keyFollower)
        print("The AdventCoint MD5 hash for ", end = "")
        print(secretKey, end = "")
        print(" is: ", end = "")
        print(hashedKey.hexdigest())
        break
    else:
        keyFollower += 1

while True:
    inputKey = secretKey + str(keyFollower)
    hashedKey = hashlib.md5(inputKey.encode())

    if hashedKey.hexdigest()[0:6] == "000000":
        print("The lowest positive number leading to the MD5 hash starting with at least 6 zeroes is: ", end = "")
        print(keyFollower)
        print("The AdventCoint MD5 hash for ", end = "")
        print(secretKey, end = "")
        print(" is: ", end = "")
        print(hashedKey.hexdigest())
        break
    else:
        keyFollower += 1
        

File: 2015/Day04_2015.java
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;

public class Day04_2015 {
    public static void main(String[] args) throws NoSuchAlgorithmException
    {
        String secretKey = "ckczppom";
        int keyFollower = 1;

        // Create MD5 digest instance
        MessageDigest md5 = MessageDigest.getInstance("MD5");

        while (true)
        {
            String inputKey = secretKey + String.valueOf(keyFollower);
            byte[] inputBytes = inputKey.getBytes(StandardCharsets.UTF_8);

            byte[] outputBytes = md5.digest(inputBytes);

            // Don't actually need to convert to hex. Just check the first 3 byte values.
            if (outputBytes[0] == 0 && outputBytes[1] == 0 && (outputBytes[2] >= 0 && outputBytes[2] < 16))
            {
                System.out.println("The lowest positive number leading to the MD5 hash starting with at least 5 zeroes is: " + keyFollower);
                break;
            }
            else
                keyFollower++;
        }

        while (true)
        {
            String inputKey = secretKey + String.valueOf(keyFollower);
            byte[] inputBytes = inputKey.getBytes(StandardCharsets.UTF_8);

            byte[] outputBytes = md5.digest(inputBytes);

            // Don't actually need to convert to hex. Just check the first 3 byte values.
            if (outputBytes[0] == 0 && outputBytes[1] == 0 && outputBytes[2] == 0)
            {
                System.out.println("The lowest positive number leading to the MD5 hash starting with at least 6 zeroes is: " + keyFollower);
                break;
            }
            else
                keyFollower++;
        }
    }
}